lib/package_json.rb

```diff
@@ -79,16 +79,19 @@ def determine_package_manager(fallback_manager)
 
     name, version = package_manager.split("@")
 
-    if name == "yarn"
-      raise Error, "a major version must be present for Yarn" if version.nil? || version.empty?
-      return :yarn_classic if version.start_with?("1")
-
-      return :yarn_berry
-    end
+    return determin_yarn_version(version) if name == "yarn"
 
     name.to_sym
   end
 
+  def determin_yarn_version(version)
+    raise Error, "a major version must be present for Yarn" if version.nil? || version.empty?
+
+    return :yarn_classic if version.start_with? "1."
```

Comment:
```suggestion
    return :yarn_classic if version == "1"
    return :yarn_classic if version.start_with? "^1."
    return :yarn_classic if version.start_with? "~1."
    return :yarn_classic if version.start_with? "1."
```

Reply:
Good point. I will update it with a regex.

Since `packageManager` entry is still experimental, is it safe to assume we get full semantic version (`x.y.z`) and not something like `1.x` or `1.5`?

If we should check for valid version string, it should be done earlier.

Reply:
We should be caring about the version at little as possible as we're explicitly expecting ensuring the right exact version of the package manager is handled elsewhere; the only reason we care at all is because of Yarn Berry vs Classic and all we need for that is to know if the major version is 1 (or not).

Happy for a regex to be used but make sure you add a test for each specific case since coverage won't catch if you miss any like it would with individual branches.

Reply:
I have updated the PR, and I decided to test the method directly rather than testing it with `read` method.

As a result, the current tests are written using `send` method for the private `determin_yarn_version`.

Should we extract such functionalities out of the class and put them in a module?

Reply:
No because this is an implementation detail of `PackageJson` so you shouldn't be using `send` at all - we've already got tests for this behaviour that don't use send, and `PackageJson` is very quick so there shouldn't be a reason to need to test this private method directly.

Instead reimplement your tests based off this one:

```
      it "supports yarn classic" do
        with_package_json_file({ "version" => "1.0.0", "packageManager" => "yarn@1.2.3" }) do
          package_json = described_class.read

          expect(package_json.manager).to be_a PackageJson::Managers::YarnClassicLike
        end
      end
```

Reply:
Isn't it better to do unit testing in such cases? If the whole logic is encapsulated in a method, I think testing it as a unit is better. That way, we can extract the regex test from tests for `PackageJson.read`, make it cleaner, and focus only on what should be tested there.

Reply:
In isolation you're right but this is encapsulated within a class because it's an implementation detail of that class which isn't needed elsewhere nor needs to be part of the public API of this gem; if we switched to determining the yarn version another way (such as via the `yarn.lock`) then the tests I'm asking you to write would barely change (technically they'd need a line to generate the `yarn.lock`, but they'd not change in a _significant_ way).

This isn't a complicated bit of logic nor are the tests I proposed expensive so we don't need to overcook this - testing through `PackageJson.read` is fine.

Reply:
It makes sense 👍🏾 
So I will move the tests to the `read` method.

By the way, I didn't cover versions with `>` and `>=`, assuming this means we go with yarn berry in such cases. But I should add support for `=1`.

Any thought?

Reply:
We might as well support `=` even though I'm doubtful it'll get used (I've literally _never_ seen it 😅), but `>` and `>=` don't have an upper bounds so using Yarn Berry for those is correct; I'm not that worried about not having tests for those but happy if you want to chuck them in